### PR TITLE
Make modal Python 3.12 compatible

### LINKED ISSUE #1056: Support for python 3.12
I tried to use modal with the newest python version (3.12) but I kept getting this error:

```
➜  modal
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniconda/base/envs/llms/bin/modal", line 5, in <module>
    from modal.__main__ import main
  File "/opt/homebrew/Caskroom/miniconda/base/envs/llms/lib/python3.12/site-packages/modal/__init__.py", line 5, in <module>
    from .app import container_app, is_local
  File "/opt/homebrew/Caskroom/miniconda/base/envs/llms/lib/python3.12/site-packages/modal/app.py", line 17, in <module>
    from .object import _Object
  File "/opt/homebrew/Caskroom/miniconda/base/envs/llms/lib/python3.12/site-packages/modal/object.py", line 23, in <module>
    _BLOCKING_O = synchronize_api(O)
                  ^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/llms/lib/python3.12/site-packages/modal_utils/async_utils.py", line 31, in synchronize_api
    return synchronizer.create_blocking(obj, blocking_name, target_module=target_module)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/llms/lib/python3.12/site-packages/synchronicity/synchronizer.py", line 687, in create_blocking
    wrapped = self._wrap(obj, Interface.BLOCKING, name, target_module=target_module)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/llms/lib/python3.12/site-packages/synchronicity/synchronizer.py", line 623, in _wrap
    if self._wrapped_attr not in obj.__dict__:

```
    
  When I downgraded python to `3.11.5` it worked fine!.

### PR DESCRIPTION
**Pull request by mwaskom**
## Describe your changes

Adds Python 3.12 to Modal's CI builds and updates package metadata.

Follow-up on https://github.com/modal-labs/modal-client/pull/1051; this was failing a few months ago as not all dependencies were ready. Trying again as it's been a while.

- Resolves MOD-2091
- Fixes #1056

## Changelog

The Modal client library is now compatible with Python 3.12.

### CODE REVIEW

File: modal/__init__.py [hunk 1]
```diff
@@ -1,11 +1,4 @@
 # Copyright Modal Labs 2022
-import sys
-
-if sys.version_info >= (3, 12):
-    raise RuntimeError(
-        "This version of Modal does not support Python 3.12+. See https://github.com/modal-labs/modal-client/issues/1056"
```
Comment (ekzhang) [inline modal/__init__.py:6]: Do you think we could update this to read `3.13`?

**Review by ekzhang — COMMENTED** [review]
😁

**Comment by erikbern** [comment]
This is great – thanks for doing this!

File: modal/requirements.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # Pins Modal dependencies installed within the container runtime.
-aiohttp==3.8.3
+aiohttp==3.8.3;python_version<'3.12'
+aiohttp==3.9.1;python_version>='3.12'
```
Comment (mwaskom) [inline modal/requirements.txt:3]: Is this going to force a rebuild of all user images?

**Comment by mwaskom** [comment]
Pulling the list of tests that timeout from the CI logs as it's a little hard to find:

```
=========================== short test summary info ============================
ERROR test/cli_test.py::test_app_descriptions - Failed: Timeout >300.0s
ERROR test/cli_test.py::test_logs - Failed: Timeout >300.0s
ERROR test/client_test.py::test_client_connection_timeout - Failed: Timeout >4.0s
ERROR test/client_test.py::test_client_old_version - Failed: Timeout >300.0s
ERROR test/client_test.py::test_client_unauthenticated - Failed: Timeout >300.0s
ERROR test/client_test.py::test_client_from_env - Failed: Timeout >300.0s
ERROR test/container_test.py::test_checkpoint_and_restore_success - Failed: Timeout >300.0s
ERROR test/live_reload_test.py::test_live_reload - Failed: Timeout >300.0s
ERROR test/live_reload_test.py::test_file_changes_trigger_reloads - Failed: Timeout >300.0s
ERROR test/live_reload_test.py::test_no_change - Failed: Timeout >300.0s
ERROR test/live_reload_test.py::test_heartbeats - Failed: Timeout >300.0s
====== 406 passed, 2 skipped, 2 warnings, 11 errors in 3189.01s (0:53:09) ======
```

This is reproducible across a couple of runs and also I can reproduce it locally when running the tests (but don't have a good reproducer for any of these outside of the test framework yet).

File: modal/requirements.txt (hunk 2 above)
Comment (ekzhang) [reply modal/requirements.txt:3]: Yeah, it will

Comment (ekzhang) [reply modal/requirements.txt:3]: We have not changed requirements.txt for a long time, since before we were concerned about rebuilds. Maybe we'll need to figure that out next 😬

Comment (mwaskom) [reply modal/requirements.txt:3]: Hum, I wonder if we could split out the base requirements file by Python version 🤔

File: modal/__init__.py (hunk 1 above)
Comment (mwaskom) [reply modal/__init__.py:6]: Good idea

File: modal/requirements.txt (hunk 2 above)
Comment (mwaskom) [reply modal/requirements.txt]: OK gave that a shot

File: setup.cfg [hunk 3]
```diff
@@ -28,11 +28,12 @@ install_requires =
...
     click>=8.1.0
     fastapi
-    grpclib==0.4.3
+    grpclib==0.4.7
     importlib_metadata>=3.6.0
     protobuf>=3.19,<5.0,!=4.24.0
     rich>=12.0.0
-    synchronicity~=0.5.3
+    # TODO revert after testing synchronicity~=0.5.3
+    synchronicity @ git+https://github.com/modal-labs/synchronicity.git@michael/ci-python-versions
```
Comment (mwaskom) [inline setup.cfg:36]: TODO: Land the change in synchronicity, cut a release, then revert this change and bump the synchronicity pin.

**Review by ekzhang — APPROVED** [review]
I think this looks good to me. Thank you for documenting all of the changes and ensuring that the requirements path is consistent!
